hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,266 @@
+  private final List<ByteBuffer> buffers = new ArrayList<>();
+  /** Is this a duplicated buffer? */
+  private final boolean isDuplicated;
+
+  IncrementalChunkBuffer(int limit, int increment, boolean isDuplicated) {
+    Preconditions.checkArgument(limit >= 0);
+    Preconditions.checkArgument(increment > 0);
+    this.limit = limit;
+    this.increment = increment;
+    this.isDuplicated = isDuplicated;
+  }
+
+  /** @return the capacity for the buffer at the given index. */
+  private int getBufferCapacityAtIndex(int i) {
+    if (limit == 0) {
+      return 0;
+    } else if (i < limit/increment) {
+      return increment;
+    }
+    return limit % increment;
+  }
+
+  private void assertInt(int expected, int computed, String name, int i) {
+    ChunkBuffer.assertInt(expected, computed,
+        () -> this + ": Unexpected " + name + " at index " + i);
+  }
+
+  /** @return the i-th buffer if it exists; otherwise, return null. */
+  private ByteBuffer getAtIndex(int i) {
+    final ByteBuffer ith = i < buffers.size() ? buffers.get(i) : null;
+    if (ith != null) {
+      if (!isDuplicated) {
+        assertInt(getBufferCapacityAtIndex(i), ith.capacity(), "capacity", i);
+      } else {
+        final int limitIndex = limit / increment;
+        if (i < limitIndex) {
+          assertInt(increment, ith.capacity(), "capacity", i);
+        } else if (i == limitIndex) {
+          assertInt(getBufferCapacityAtIndex(i), ith.limit(), "capacity", i);
+        } else {
+          assertInt(0, ith.limit(), "capacity", i);
+        }
+      }
+    }
+    return ith;
+  }
+
+  /** @return the i-th buffer. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtIndex(int index) {
+    final int limitIndex = limit / increment;
+    // never allocate over limit
+    if (limit % increment == 0) {
+      Preconditions.checkArgument(index < limitIndex);
+    } else {
+      Preconditions.checkArgument(index <= limitIndex);
+    }
+
+    int size = buffers.size();
+    if (index < size) {
+      return getAtIndex(index);
+    }
+
+    // allocate upto the given index
+    ByteBuffer b = null;
+    for (; size <= index; size++) {
+      b = ByteBuffer.allocate(getBufferCapacityAtIndex(size));
+      buffers.add(b);
+    }
+    return b;
+  }
+
+  /** @return the buffer containing the position. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtPosition(int position) {
+    Preconditions.checkArgument(position < limit);
+    final int i = position / increment;
+    final ByteBuffer ith = getAndAllocateAtIndex(i);
+    assertInt(position%increment, ith.position(), "position", i);
+    return ith;
+  }
+
+  @Override
+  public int position() {
+   // The buffers list must be in the following orders:
+   // full buffers, buffer containing the position, empty buffers, null buffers
+    int i = 0;
+    ByteBuffer ith = null;
+    // buffers must be full
+    for (; i < buffers.size(); i++) {
```

Comment:
Wouldn't it be easier to track this location? instead of iterating ?

Reply:
We do not have a position field in this class so that we don't have to worry about synchronizing the position field with each positions in each ByteBuffers in the list.